### PR TITLE
Blog post about Angular and Quarkus development.

### PR DESCRIPTION
**Pull request by kabir**

### CODE REVIEW

File: _posts/2019-05-23-quarkus-and-angular-ui-development-mode.adoc [hunk 1]
```diff
@@ -0,0 +1,391 @@
...
+
+First off scaffold the project (substitute the '0.15.0' with the latest and greatest Quarkus release version):
+
+```
+$mvn io.quarkus:quarkus-maven-plugin:0.15.0:create \
+    -DprojectGroupId=org.kabir.quarkus \
+    -DprojectArtifactId=blog-quarkus-ui-development \
+    -DprojectVersion=0.1.0 \
+    -DclassName="org.kabir.quarkus.ui.SampleResource" \
+    -Dextensions=io.quarkus:quarkus-resteasy,io.quarkus:quarkus-resteasy-jsonb
```
Comment (rsvoboda) [inline _posts/2019-05-23-quarkus-and-angular-ui-development-mode.adoc]: ` -Dextensions=io.quarkus:quarkus-resteasy-jsonb` is sufficient

quarkus-resteasy is already added in template - https://github.com/quarkusio/quarkus/blob/master/devtools/common/src/main/resources/templates/basic-rest/java/pom-template.ftl#L31

Right now user will see warnings like:
```
[WARNING] 'dependencies.dependency.(groupId:artifactId:type:classifier)' must be unique:
   io.quarkus:quarkus-resteasy:jar -> duplicate declaration of version (?) @ line 46, column 17
```

File: _posts/2019-05-23-quarkus-and-angular-ui-development-mode.adoc [hunk 2]
```diff
@@ -0,0 +1,391 @@
...
+----
+
+
+All this does is that when we go to the root of the application, we will go to DefaultComponent. `DefaultComponent` has routerLinks to `OtherComponent` and `RestComponent` (for routerLinks Angular does not hit the back-end), and a normal link to `/servlet/make-external-call`.
+
+If we go to `/other`, we end up in `OtherComponent` which just has a link back to `DefaultComponent`.
+
+If we go to `/rest`, we end up in `RestComponent` which displays data from the `SampleResource` we saw above and also has a link back to `DefaultComponent`.
+
+Finally if we go to `/clientCallback` (which is triggered via `/servlet/callback`  in `/SampleServlet` we end up in `ClientCallbackComponent`.
```
Comment (rsvoboda) [inline _posts/2019-05-23-quarkus-and-angular-ui-development-mode.adoc]: `/SampleServlet` => `SampleServlet`

File: _posts/2019-05-23-quarkus-and-angular-ui-development-mode.adoc [hunk 3]
```diff
@@ -0,0 +1,391 @@
...
+
+====== Illustrating the problem
+Now package and start the application by running:
+[source,bash]
+----
+$mvn package quarkus:dev -Dui.dev
+----
+If you go to http://localhost:8080 you will get a page with links to `Other`, `Rest` and `Default`. Click on the `Other` and `Rest` ones, and it should all work.
+
+However while in the `Other` component, so that the address in the browser is http://localhost:8080/other, if you try to refresh the page you will end up with the following error message:
```
Comment (rsvoboda) [inline _posts/2019-05-23-quarkus-and-angular-ui-development-mode.adoc]: This message look a bit strange. What about following version?

However while in the `Other` component (http://localhost:8080/other), if you try to refresh the page you will end up with the following error message:

File: _posts/2019-05-23-quarkus-and-angular-ui-development-mode.adoc [hunk 4]
```diff
@@ -0,0 +1,391 @@
...
+        HttpServletResponse response = (HttpServletResponse) res;
+
+        if (isIndexHtmlRequest(request)) {
+            request.getRequestDispatcher("/").forward(request, response);
+        } else {
+            chain.doFilter(request, response);
+        }
+    }
+
+    private boolean isIndexHtmlRequest(HttpServletRequest request) {
```
Comment (rsvoboda) [inline _posts/2019-05-23-quarkus-and-angular-ui-development-mode.adoc]: I would change the method name, it's confusing.

`assumeAngularRoute()`  or `isAngularRoute` sound better to me

File: _posts/2019-05-23-quarkus-and-angular-ui-development-mode.adoc [hunk 5]
```diff
@@ -0,0 +1,391 @@
...
+layout: post
+title: Quarkus and Angular UI Development
+date: 2019-05-23 14:00:00.000 -0600
+tags: quarkus ui development
+author: kkhan
+---
+
+== Quarkus and Angular UI Development
+
+In this blog post we will look at how to take advantage of the respective development modes of both Quarkus and Angular CLI and see how we can develop a zero turnaround web application backed by a RESTful API on Quarkus. While I am using Angular, the concepts should be the same for other web application frameworks.
```
Comment (rsvoboda) [inline _posts/2019-05-23-quarkus-and-angular-ui-development-mode.adoc]: >  the concepts should be the same for other web application frameworks.

Mention you looked at React and Vue and they support proxying API requests.

So the reader familiar with above two doesn't give up reading when not familiar with Angular

File: _posts/2019-05-23-quarkus-and-angular-ui-development-mode.adoc [hunk 6]
```diff
@@ -0,0 +1,391 @@
...
+
+Now let's try it out. You need two terminals.
+
+In the first terminal run:
+[source,bash]
+----
+$mvn clean
+$mvn package quarkus:dev
+----
+to start the Quarkus application. Note how we did not pass in `-Dui.dev` so we will no longer bundle the web application which saves us a significant amount of time. If you go to http://localhost:8080 you will get an error since we have no web application this time.
```
Comment (rsvoboda) [inline _posts/2019-05-23-quarkus-and-angular-ui-development-mode.adoc]: `you will get an error`

That's true if you remove `webapp/dist/webapp` first.

File: _posts/2019-05-23-quarkus-and-angular-ui-development-mode.adoc [hunk 7]
```diff
@@ -0,0 +1,391 @@
...
+
+Now go to http://locahost:4200 and see everything working smoothly.
+
+==== Conclusion
+We have seen how to package an Angular and Quarkus application, and tweaks needed to make it behave in a development environment. The tweaks needed are quite small, and offer great developer productivity when working on your application. You basically just fire up Quarkus with the system property we added (`mvn package quarkus:dev -Dui.proxy=true`), and we do the same for Angular to run it in proxy mode (`yarn proxy`). Now we can just forget about it and modify both back-end and front-end code and see changes happen on the next browser refresh. There is no need to repackage and restart the application with every change done in either place.
+
+Finally, to run this in production you need to run:
+[source,bash]
+----
+$mvn package -Dui
```
Comment (rsvoboda) [inline _posts/2019-05-23-quarkus-and-angular-ui-development-mode.adoc]: Add section about native image, this is very important part of Quarkus story.

`mvn package -Dui -Pnative` works fine and you see awesome 
```
Quarkus 0.15.0 started in 0.007s. Listening on: http://[::]:8080
```

**Comment by rsvoboda** [comment]
@gsmet / @stuartwdouglas / @emmanuelbernard 

Can you look into this blog post ?

It could possibly go out together with 0.16.0 release.

**Comment by kabir** [comment]
I will address @rsvoboda 's comments, but as it is a slight pain to retag all the code when I make changes I will wait for further comments from people. Or if someone lets me know it looks ok once I've incorporated those changes, I can do so then

**Comment by emmanuelbernard** [comment]
General comments are the front end is not my forte.
To me, the blog feels like you went through a lot of pains and hitting snatches but by luck recovered every time. Not sure that's the feeling we want to give. It looks like you want to explain the various points and then find a solution but I had this negative reading.

Have you tried and explored the idea of wording it with the following sections:

1. what we aim to achieve
2. ninja survival guide jsut ot get started (for the tl;dr;
3. a diagram explaining the various problems
4. the detailed solution that overcomes points in 3

I don't know if it would work best so take it as a proposal.

The other points are more technical:

1. servlets are a bit passé so having to define some is not awesome but a better solution will have to wait for the router API I suppose CC @stuartwdouglas 
2. you have a lot of "stop quarkus and start it again which is not a nice experience. I wonder if that was mandatory or if that simply was due to you wanting to do step by step and therefore having these stops to update your poms and co?

**Comment by kabir** [comment]
The stop and start is to do all this step by step, yes. So it leads up to being able to do everything without needing to stop and start it. 

To have something to demo (steps 1-3), it became quite lengthy for something which is not really that interesting, and which is where the bulk of the stop/start stuff is. I could push steps 1-3 into the example code repositories README and just refer to that from the blog. Then I can: 
* briefly run through the three problems I have seen
* Address them with steps 4-6

Hopefully, once the whole thing is shorter I can make clearer what is actually needed (aka your ninja survival guide)

**Comment by rsvoboda** [comment]
> I could push steps 1-3 into the example code repositories README

I would push full blog post to README, including steps 4-6

**Comment by kabir** [comment]
Replaced by #223

**Comment by rsvoboda** [comment]
Closing as this PR is invalid, followup one is merged.

**Pull request closed without merging**